### PR TITLE
Improve equivalence checking

### PR DESCRIPTION
**Pull request by drganam**
Improvements to equivalence checking phase, including: function call matching, conditional equivalence (--norm option), and user-provided test generators (mkTest annotation)

### CODE REVIEW

**Comment by vkuncak** [comment]
@mario-bucev @samarion : Any chance you could take a quick look? Equivalence checking is already in master; this PR improves it. Most changes are localized in that phase and I we do not necessarily need a detailed review of that one; I suggest you focus on any possible interference (my quick check indicates this is minimal).

File: core/src/main/scala/stainless/extraction/trace/Trace.scala [hunk 1]
```diff
@@ -37,58 +58,270 @@ class Trace(override val s: Trees, override val t: termination.Trees)
...
+        val f2Calls = getFunCalls(fd2).filter(!_.flags.exists(_.name == "library"))
+
+        val pairs = f1Calls zip f1Calls.map(m => f2Calls.find(f => m != f && checkArgs(m, f) && f.id.name == m.id.name).orElse(f2Calls.find(f => m != f && checkArgs(m, f))))
+
+        val validpairs = pairs.filter(elem => elem._2 != None)
+
+        validpairs.map(elem => (elem._1, elem._2) match {
+          case (m, Some(f)) =>
+            (equivalenceCheck(m, f, true), m, f)
+        })
```
Comment (mario-bucev) [inline core/src/main/scala/stainless/extraction/trace/Trace.scala:222]: Minor suggestion
```suggestion
        pairs.collect {
          case (m, Some(f)) =>
            (equivalenceCheck(m, f, true), m, f)
        }
```

**Review by mario-bucev — COMMENTED** [review]
Looks good to me :)

File: core/src/main/scala/stainless/extraction/trace/Trace.scala (hunk 1 above)
Comment (drganam) [reply core/src/main/scala/stainless/extraction/trace/Trace.scala]: Updated and squashed, thanks!

**Comment by mario-bucev** [comment]
The CI is not happy, let us spank it a bit

**Comment by drganam** [comment]
Now it's back to happy

**Comment by mario-bucev** [comment]
@vkuncak Can we merge?

**Pull request merged**